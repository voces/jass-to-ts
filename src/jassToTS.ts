import jassToAst, {
	ArrayRef,
	BinaryOp,
	Call,
	Comment,
	Debug,
	Else,
	ElseIf,
	EmptyLine,
	ExitWhen,
	FourCC,
	FuncRef,
	Globals,
	IfThenElse,
	JASSFunction,
	JASSSet,
	List,
	Loop,
	Name,
	Native,
	Node,
	Param,
	Parens,
	Program,
	Return,
	SingleProp,
	Statements,
	Type,
	UnaryOp,
	Variable,
} from "jass-to-ast";

const jassTypeToTypeScriptType = (type: Name): string =>
	type.replace("real", "number").replace("integer", "number");

let isVoid = false;

const astToTS = (
	ast: Node | List<Node> | SingleProp | string | boolean,
): string => {
	if (ast === undefined || ast === null) return "null";

	if (typeof ast === "number" || typeof ast === "boolean")
		return ast.toString();

	if (ast instanceof ArrayRef) return `${ast.name}[ ${astToTS(ast.prop)} ]`;

	if (ast instanceof BinaryOp) {
		const operator = ast.operator
			.replace("or", "||")
			.replace("and", "&&")
			.replace("==", "===")
			.replace("!=", "!==");
		return `${astToTS(ast.left)} ${operator} ${astToTS(ast.right)}`;
	}

	if (ast instanceof Call)
		return `${ast.name}(${
			ast.args ? `${[...ast.args].map(astToTS).join(", ")}` : ""
		})`;

	if (ast instanceof Comment)
		return ast.includes("\n") ? "/*" + ast + "*/" : "//" + ast;

	if (ast instanceof Else) return `else {\n\n${astToTS(ast.statements)}\n\n}`;

	if (ast instanceof ElseIf)
		return `else if ( ${astToTS(ast.condition)} ) {\n\n${astToTS(
			ast.statements,
		)}\n\n}`;

	if (ast instanceof EmptyLine) return "";

	if (ast instanceof ExitWhen) return `if (${astToTS(ast.data[0])}) break;`;

	if (ast instanceof FourCC) return `FourCC("${ast}")`;

	if (ast instanceof FuncRef) return ast.data;

	if (ast instanceof Globals)
		return [
			ast.comment,
			astToTS(ast.globals)
				.split("\n")
				.map((v) => v.slice(1))
				.join("\n"),
			ast.endComment,
		]
			.filter(Boolean)
			.join("\n");

	if (ast instanceof IfThenElse)
		return `if (${astToTS(ast.condition)}) {\n${astToTS(ast.then)}\n}${
			ast.elses ? " " + ast.elses.map(astToTS).join(" ") : ""
		}`;

	if (ast instanceof JASSFunction) {
		if (ast.returns === undefined) isVoid = true;

		const bodyContent = ast.statements ? astToTS(ast.statements) : "";
		const body = bodyContent ? `\n${bodyContent}\n` : "";
		const r = `const ${ast.name} = (${
			ast.params ? `${[...ast.params].map(astToTS).join(", ")}` : ""
		}): ${jassTypeToTypeScriptType(ast.returns || "void")} => {${body}};`;
		isVoid = false;
		return r;
	}

	if (ast instanceof JASSSet) {
		const name = ast.name.replace("this", "_this");
		const prop = ast.prop !== undefined ? `[ ${astToTS(ast.prop)} ]` : "";
		return `${name}${prop} = ${astToTS(ast.value)};`;
	}

	if (ast instanceof Loop)
		return `while (true) {\n${[...ast.statements]
			.map(astToTS)
			.join("\n")
			.split("\n")
			.map((v) => (v ? "\t" + v : v))
			.join("\n")}\n}`;

	if (ast instanceof Name) return ast.replace("this", "_this");

	if (ast instanceof Native) if (ast instanceof Type) return "";

	if (ast instanceof Param)
		return `${ast.name.replace(
			"this",
			"_this",
		)}: ${jassTypeToTypeScriptType(ast.type)}`;

	if (ast instanceof Parens) return `( ${astToTS(ast.data[0])} )`;

	if (ast instanceof Program) return [...ast].map(astToTS).join("\n");

	if (ast instanceof Return)
		if (!isVoid) return `return ${astToTS(ast.data[0])};`;
		else return "return;";

	if (ast instanceof Statements)
		return [...ast]
			.map(astToTS)
			.join("\n")
			.split("\n")
			.map((v) => (v ? "\t" + v : v))
			.join("\n");

	if (ast instanceof UnaryOp)
		return `${ast.operator.replace("not", "!")} ${astToTS(ast.expr)}`;

	if (ast instanceof Variable) {
		const decl = ast.constant ? "const" : "let";
		const name = ast.name.replace("this", "_this");
		const baseType = jassTypeToTypeScriptType(ast.type);
		const type = ast.array ? `Array<${baseType}>` : baseType;
		const hasInferableValue = "value" in ast;
		const value = hasInferableValue
			? " = " + astToTS(ast.value)
			: ast.array
			? " = []"
			: "";

		return `${decl} ${name}${
			hasInferableValue ? "" : `: ${type}`
		}${value};`;
	}

	if (ast instanceof String || typeof ast === "string") return `"${ast}"`;

	if (ast instanceof Debug) return "";

	throw new Error("Unknown AST node " + ast.constructor.name);
};

export const jassToTS = (jass: string): string => astToTS(jassToAst(jass));
