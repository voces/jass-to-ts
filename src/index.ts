#!/usr/bin/env node

import fs from "fs";

import { jassToTS } from "./jassToTS.js";

const filePath = process.argv[2];

if (!filePath) {
	console.error("You must specify a jass file to convert");
	console.error("Usage: jassToTS path/to/war3map.j");
	process.exit(1);
}

fs.readFile(filePath, "utf-8", (err, res) => {
	if (err) {
		console.error(err.message);
		process.exit(1);
	}

	// eslint-disable-next-line no-console
	console.log(jassToTS(res.replace(/\r/g, "")));
});
