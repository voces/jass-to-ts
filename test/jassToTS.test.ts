/* eslint @typescript-eslint/explicit-function-return-type: 0 */

import { jassToTS } from "../src/jassToTS";

const trimEmptyLines = (str: string) => {
	const lines = str.split("\n");

	let start = 0;
	while (start < lines.length && lines[start].trim() === "") start++;

	if (start === lines.length) return "";

	let end = lines.length - 1;
	while (end >= 0 && lines[end].trim() === "") end--;

	if (end === -1) return "";

	return lines.slice(start, end + 1).join("\n");
};

const trim = (str: string) => {
	const lines = trimEmptyLines(str);
	const match = lines.match(/\s+/);
	if (match) {
		const indent = match[0];
		const regex = new RegExp("^" + indent);
		return lines
			.split("\n")
			.map((line) => line.replace(regex, ""))
			.join("\n");
	}

	return lines;
};

const testProgram = (name: string, jass: string) =>
	test(name, () => expect(jassToTS(trim(jass))).toMatchSnapshot());

testProgram(
	"globals",
	`
		globals

			// my numbers
			constant real testNumber = 7.12

			// my strings
			string testString = "myString"

			// stress tests
			unit array myUnits //tehe

		endglobals
	`,
);

testProgram(
	"simple",
	`
		globals
			//comment
			unit udg_blah = null
		endglobals

		function myFunc takes integer i, boolean b returns nothing
			local unit someUnit = GetSummonedUnit()
			if unit == null then
				set unit = CreateUnit( Player(0), 'hfoo', 0, 0, 0 )
				set udg_blah = unit
			endif
			if R2I(GetUnitX(unit)) < i then
				call ShowUnit(unit, b)
			endif
			loop
				exitwhen i > 10
				set i = i + 1
			endloop
		endfunction
	`,
);

testProgram(
	"empty function",
	`
		function InitGlobals takes nothing returns nothing
		endfunction
	`,
);

testProgram(
	"debug statements",
	`
		function MyFunc takes nothing returns nothing
			debug call BJDebugMsg("hello, world!")
		endfunction
	`,
);

testProgram(
	"multiline comments",
	`
		/**
		 * Blah blah blah
		 * @param arg1 Blah blah blah
		 */
		function MyFunc takes string arg1 returns nothing
		endfunction
	`,
);

testProgram(
	"variables",
	`
		globals
			constant real bj_PI = 3.14159
			constant real bj_RADTODEG = 180.0 / bj_PI
		endglobals
	`,
);
