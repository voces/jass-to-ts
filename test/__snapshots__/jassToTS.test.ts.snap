// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`debug statements 1`] = `"const MyFunc = (): void => {};"`;

exports[`empty function 1`] = `"const InitGlobals = (): void => {};"`;

exports[`globals 1`] = `
"
// my numbers
const testNumber = 7.12;

// my strings
let testString = \\"myString\\";

// stress tests
let myUnits: Array<unit> = [];
"
`;

exports[`multiline comments 1`] = `
"/**
 * Blah blah blah
 * @param arg1 Blah blah blah
 */

const MyFunc = (arg1: string): void => {};"
`;

exports[`simple 1`] = `
"//comment
let udg_blah: unit;

const myFunc = (i: number, b: boolean): void => {
	let someUnit = GetSummonedUnit();
	if (unit === null) {
		unit = CreateUnit(Player(0), FourCC(\\"hfoo\\"), 0, 0, 0);
		udg_blah = unit;
	}
	if (R2I(GetUnitX(unit)) < i) {
		ShowUnit(unit, b)
	}
	while (true) {
		if (i > 10) break;
		i = i + 1;
	}
};"
`;

exports[`variables 1`] = `
"const bj_PI = 3.14159;
const bj_RADTODEG = 180 / bj_PI;"
`;
